File: src/engine/game.ts
/**
 * Game state machine — orchestrates phase transitions and action dispatch.
 *
 * The Game class is the central controller. It holds the current GameState
 * and provides methods to transition between phases.
 *
 * The engine calls Rules layer functions for validation; it does NOT
 * contain rule logic itself.
 */

import { Card, Suit, Rank, sortHand } from '../core/card';
import { Phase, PlayType, type PlayAction, type GameState, type BidAction, type StirAction, type GameSettings, type PlayerState } from '../core/types';
import { PLAYER_COUNT, HUMAN_PLAYER_INDEX } from '../core/constants';
import { getTeamIndex, nextPlayer, clockwiseDistance } from '../core/constants';
import {
  createInitialState,
  dealCards,
  recordBid,
  setDeclarer,
  recordStir,
  pickupBottomCards,
  discardCards,
  playCards,
  advanceRound,
  serializeForAI,
} from './state';
import { calculateScore, isGameOver } from './scoring';
import { isValidBid, getValidBidLevels, isBiddingOver, getWinningBid, isValidStir, getValidStirOptions } from '../rules/bidding';
import { getLegalPlays } from '../rules/validator';

export type GameEventCallback = (state: GameState) => void;

export class Game {
  state: GameState;
  private listeners: Set<GameEventCallback> = new Set();
  private static STORAGE_KEY = 'tractor-game-state';

  constructor(settings?: Partial<GameSettings>) {
    // Try restoring saved state first
    const saved = Game.loadState();
    if (saved) {
      this.state = { ...saved, settings: { ...saved.settings, ...settings } };
    } else {
      this.state = createInitialState(settings);
    }
  }

  /** Subscribe to state changes. */
  onChange(cb: GameEventCallback): () => void {
    this.listeners.add(cb);
    return () => this.listeners.delete(cb);
  }

  private emit(): void {
    // Sort player hands before notifying
    this.state = {
      ...this.state,
      players: this.state.players.map(p => ({
        ...p,
        hand: sortHand(p.hand),
      })) as [PlayerState, PlayerState, PlayerState, PlayerState],
    };

    // Persist to localStorage
    Game.saveState(this.state);

    for (const cb of this.listeners) {
      cb(this.state);
    }
  }

  /** Save game state to localStorage. */
  private static saveState(state: GameState): void {
    try {
      localStorage.setItem(Game.STORAGE_KEY, JSON.stringify(state));
    } catch {
      // localStorage full or unavailable — just skip
    }
  }

  /** Load game state from localStorage. Returns null if none or corrupt. */
  private static loadState(): GameState | null {
    try {
      const raw = localStorage.getItem(Game.STORAGE_KEY);
      if (!raw) return null;
      const parsed = JSON.parse(raw);
      // Basic sanity check
      if (parsed && parsed.phase && parsed.players && parsed.players.length === 4) {
        return parsed as GameState;
      }
    } catch {
      // Corrupt data — ignore
    }
    return null;
  }

  /** Clear saved game state. */
  static clearSavedState(): void {
    localStorage.removeItem(Game.STORAGE_KEY);
  }

  // ---- Game Flow ----

  /** Start a new game. */
  startNewGame(): void {
    Game.clearSavedState();
    this.state = createInitialState(this.state.settings);
    this.emit();
  }

  /** Deal cards and begin bidding. */
  startRound(): void {
    this.state = dealCards(this.state);
    this.emit();
  }

  // ---- Bidding ----

  /** Submit a bid for the current player. */
  submitBid(playerIndex: number, level: Rank | null, pass: boolean): boolean {
    if (this.state.phase !== Phase.BIDDING) return false;
    if (playerIndex !== this.state.currentPlayerIndex) return false;

    const highestBid = getWinningBid(this.state.biddingHistory);
    const highestLevel = highestBid?.level ?? null;

    if (!isValidBid(level, pass, highestLevel, this.state.currentLevel, this.state.biddingHistory.length === 0)) {
      return false;
    }

    const bid: BidAction = { playerIndex, level, pass };
    this.state = recordBid(this.state, bid);

    // Check if bidding is over
    if (isBiddingOver(this.state.biddingHistory, PLAYER_COUNT)) {
      const winner = getWinningBid(this.state.biddingHistory);
      if (winner && winner.level) {
        // Winner must choose trump suit
        // If human: wait for UI input -> handled by setTrumpAfterBid
        // If AI: handled externally
        this.emit();
      } else {
        // No one bid — redeal
        this.startRound();
      }
    } else {
      this.emit();
    }

    return true;
  }

  /** Set trump suit after winning the bid. */
  setTrumpAfterBid(playerIndex: number, trumpSuit: Suit): boolean {
    const winner = getWinningBid(this.state.biddingHistory);
    if (!winner || winner.playerIndex !== playerIndex) return false;
    if (this.state.phase !== Phase.BIDDING) return false;

    const bidLevel = winner.level!;
    this.state = setDeclarer(this.state, playerIndex, trumpSuit, bidLevel);
    this.emit();
    return true;
  }

  /** Get valid bid levels for the current player. */
  getValidBids(): Rank[] {
    const highestBid = getWinningBid(this.state.biddingHistory);
    return getValidBidLevels(highestBid?.level ?? null, this.state.currentLevel);
  }

  // ---- Stirring (炒地皮) ----

  /** Submit a stir or pass. */
  submitStir(playerIndex: number, stir: StirAction | null): boolean {
    if (this.state.phase !== Phase.STIRRING) return false;
    if (playerIndex !== this.state.currentPlayerIndex) return false;

    if (stir === null) {
      // Pass on stirring
      // Move to next player
      this.state = {
        ...this.state,
        currentPlayerIndex: nextPlayer(playerIndex),
      };

      // Check if stirring round is over (all passed since last stir)
      const lastStirIdx = this.state.stirHistory.length > 0
        ? this.state.stirHistory[this.state.stirHistory.length - 1].playerIndex
        : getWinningBid(this.state.biddingHistory)!.playerIndex;

      // Count clockwise steps from lastStirIdx to current player
      const playersSinceStir = clockwiseDistance(lastStirIdx, this.state.currentPlayerIndex);
      if (playersSinceStir === 0) {
        // Everyone passed — stirring over, move to exchange
        this.state = pickupBottomCards(this.state);
      }
    } else {
      if (!isValidStir(stir, this.state.trumpSuit!, this.state.trumpRank, this.state.stirHistory, playerIndex)) {
        return false;
      }
      this.state = recordStir(this.state, stir);
    }

    this.emit();
    return true;
  }

  /** Complete stirring phase (when all pass) and begin exchange. */
  completeStirring(): void {
    if (this.state.phase !== Phase.STIRRING) return;
    this.state = pickupBottomCards(this.state);
    this.emit();
  }

  /** Get valid stir options for the current player. */
  getValidStirs(): StirAction[] {
    return getValidStirOptions(
      this.state.trumpSuit!,
      this.state.trumpRank,
      this.state.currentPlayerIndex,
      this.state.stirHistory,
    );
  }

  // ---- Exchange (扣底) ----

  /** Declarer discards cards after picking up bottom cards. */
  submitDiscard(playerIndex: number, cards: Card[]): boolean {
    if (this.state.phase !== Phase.EXCHANGE) return false;
    if (playerIndex !== this.state.currentPlayerIndex) return false;

    const player = this.state.players[playerIndex];
    if (cards.length !== this.state.settings.bottomCardCount) return false;

    // Verify all cards are in player's hand
    const handIds = new Set(player.hand.map(c => c.id));
    for (const card of cards) {
      if (!handIds.has(card.id)) return false;
    }

    this.state = discardCards(this.state, playerIndex, cards);
    this.emit();
    return true;
  }

  // ---- Playing ----

  /** Play cards from a player's hand. */
  submitPlay(playerIndex: number, cards: Card[]): boolean {
    if (this.state.phase !== Phase.PLAYING) return false;
    if (playerIndex !== this.state.currentPlayerIndex) return false;

    // Determine the play type
    const legalPlays = this.getLegalPlays(playerIndex);
    const action = this.matchPlayAction(cards, legalPlays);

    if (!action) return false;

    this.state = playCards(this.state, playerIndex, action);
    this.emit();
    return true;
  }

  /** Get all legal plays for a player. */
  getLegalPlays(playerIndex: number): PlayAction[] {
    const player = this.state.players[playerIndex];
    const isLeading = this.state.currentTrick.every(s => s.cards === null);

    // Get lead action if following
    let leadAction: PlayAction | null = null;
    if (!isLeading) {
      const leadSlot = this.state.currentTrick.find(s => s.cards !== null);
      if (leadSlot && leadSlot.cards) {
        leadAction = {
          type: this.state.leadPlayType ?? PlayType.SINGLE,
          cards: leadSlot.cards,
        };
      }
    }

    return getLegalPlays(
      player.hand,
      this.state.currentTrick,
      this.state.trumpSuit!,
      this.state.trumpRank,
      isLeading,
      leadAction,
    );
  }

  // ---- Scoring ----

  /** Calculate scores for the round. */
  calculateRoundScore() {
    if (this.state.phase !== Phase.SCORING) throw new Error('Not in scoring phase');

    const lastTrick = this.state.trickHistory[this.state.trickHistory.length - 1];
    const lastTrickWinnerTeam = getTeamIndex(lastTrick.winnerIndex);

    return calculateScore(
      this.state.defenderPoints,
      this.state.bottomCards,
      lastTrickWinnerTeam,
      this.state.declarerTeamIndex,
      this.state.currentLevel,
    );
  }

  /** Advance to the next round. */
  nextRound(): void {
    const result = this.calculateRoundScore();

    if (isGameOver(result.nextLevel, this.state.settings.targetLevel)) {
      this.state = { ...this.state, phase: Phase.GAME_OVER };
      this.emit();
      return;
    }

    this.state = advanceRound(this.state, result.nextLevel, result.nextDeclarerTeam);
    this.emit();
  }

  // ---- Helpers ----

  /**
   * Clear the current trick display after the post-trick pause.
   * Resets currentTrick slots to empty for the next trick.
   */
  clearTrick(): void {
    this.state = {
      ...this.state,
      currentTrick: this.state.currentTrick.map(s => ({ ...s, cards: null })),
    };
    this.emit();
  }

  /** Check if it's the human player's turn. */
  isHumanTurn(): boolean {
    return this.state.currentPlayerIndex === HUMAN_PLAYER_INDEX;
  }

  /** Get the human player's hand. */
  getHumanHand(): Card[] {
    return this.state.players[HUMAN_PLAYER_INDEX].hand;
  }

  /** Get serialized state for AI prompts. */
  serializeForAI(): Record<string, unknown> {
    return serializeForAI(this.state);
  }

  /** Get state for a specific player (for AI). */
  getPlayerState(playerIndex: number) {
    const player = this.state.players[playerIndex];
    return {
      index: playerIndex,
      name: player.name,
      hand: player.hand,
      teamIndex: player.teamIndex,
      isDeclarer: player.isDeclarer,
      isHuman: player.isHuman,
    };
  }

  /** Try to match played cards to a legal play action. */
  private matchPlayAction(cards: Card[], legalPlays: PlayAction[]): PlayAction | null {
    const cardIds = new Set(cards.map(c => c.id));

    for (const play of legalPlays) {
      const playIds = new Set(play.cards.map(c => c.id));
      if (cardIds.size !== playIds.size) continue;

      let match = true;
      for (const id of cardIds) {
        if (!playIds.has(id)) {
          match = false;
          break;
        }
      }
      if (match) return play;
    }

    return null;
  }

  /** Update settings. */
  updateSettings(settings: Partial<GameSettings>): void {
    this.state = {
      ...this.state,
      settings: { ...this.state.settings, ...settings },
    };
    this.emit();
  }

  /** Begin the exchange phase after stirring ends. */
  beginExchange(): void {
    if (this.state.phase !== Phase.STIRRING) return;
    this.state = pickupBottomCards(this.state);
    this.emit();
  }
}

File: server/engine/game.py
"""Game state machine for 升级 (Shengji/Tractor).

Orchestrates phase transitions, validates actions using the rules layer,
and exposes the query/action interface that the API layer will call.
Implements AI auto-play: after a human action, the server automatically
executes AI turns until it is the human player's turn again.

Turn order is sequential (0→1→2→3→0) throughout all phases.  The Game class
manages current_player_index directly rather than delegating to state
functions (which use the clockwise seating order 0→2→3→1→0).
"""

from __future__ import annotations

from server.engine.card import Card, Rank, Suit
from server.engine.constants import HUMAN_PLAYER_INDEX, PLAYER_COUNT
from server.engine.game_state import GameState, TrickSlot
from server.engine.player_utils import get_team_index
from server.engine.scoring import ScoreResult, calculate_score, is_game_over
from server.engine.state import (
    advance_round,
    clear_trick as state_clear_trick,
    create_initial_state,
    deal_cards,
    discard_cards,
    pickup_bottom_cards,
    play_cards,
    set_declarer,
)
from server.engine.types import BidAction, Phase, PlayAction, PlayType, StirAction
from server.rules.bidding import (
    get_valid_bid_levels,
    get_valid_stir_options,
    get_winning_bid,
    is_bidding_over,
    is_valid_bid,
    is_valid_stir,
)
from server.rules.validator import get_legal_plays as _get_legal_plays


def _next_player_sequential(player_index: int) -> int:
    """Advance to next player in sequential order (0→1→2→3→0)."""
    return (player_index + 1) % PLAYER_COUNT


class Game:
    """Central game controller orchestrating phase transitions and action dispatch."""

    def __init__(self) -> None:
        self.state: GameState = create_initial_state()
        self._stir_passes: set[int] = set()

    # ---- Game Flow ----

    def start_round(self) -> None:
        """Deal cards and begin bidding."""
        self.state = deal_cards(self.state)
        self._stir_passes = set()

    # ---- Bidding ----

    def submit_bid(
        self, player_index: int, level: Rank | None, *, pass_: bool
    ) -> bool:
        """Submit a bid for a player. Returns True if accepted."""
        if self.state.phase != Phase.BIDDING:
            return False
        if player_index != self.state.current_player_index:
            return False

        highest_bid = get_winning_bid(self.state.bidding_history)
        highest_level = highest_bid.level if highest_bid else None

        if not is_valid_bid(level, pass_, highest_level, self.state.current_level):
            return False

        # Record bid manually (sequential turn order, not clockwise)
        bid = BidAction(
            player_index=player_index,
            level=level,
            pass_=pass_,
        )
        self.state = self.state.model_copy(update={
            "bidding_history": [*self.state.bidding_history, bid],
            "current_player_index": _next_player_sequential(player_index),
        })

        # Check if bidding is over
        if is_bidding_over(self.state.bidding_history, PLAYER_COUNT):
            winner = get_winning_bid(self.state.bidding_history)
            if winner is not None and winner.level is not None:
                # Winner must choose trump suit — restore their index
                self.state = self.state.model_copy(
                    update={"current_player_index": winner.player_index}
                )
            else:
                # No one bid — redeal
                self.start_round()
        elif player_index == HUMAN_PLAYER_INDEX:
            # Only auto-play AI after a human action
            self._ai_auto_play()

        return True

    def set_trump(self, player_index: int, trump_suit: Suit) -> bool:
        """Set trump suit after winning the bid. Returns True if accepted."""
        if self.state.phase != Phase.BIDDING:
            return False

        winner = get_winning_bid(self.state.bidding_history)
        if winner is None or winner.player_index != player_index:
            return False
        if winner.level is None:
            return False

        self.state = set_declarer(
            self.state, player_index, trump_suit, winner.level
        )
        # Override to sequential turn order (next player after declarer)
        self.state = self.state.model_copy(
            update={"current_player_index": _next_player_sequential(player_index)}
        )
        return True

    def get_valid_bids(self) -> list[Rank]:
        """Get valid bid levels for the current player."""
        highest_bid = get_winning_bid(self.state.bidding_history)
        return get_valid_bid_levels(
            highest_bid.level if highest_bid else None,
            self.state.current_level,
        )

    # ---- Stirring (炒地皮) ----

    def submit_stir(self, player_index: int, stir: StirAction | None) -> bool:
        """Submit a stir action or pass.

        For passes (stir=None), any player may submit regardless of turn order.
        The game tracks which players have passed and ends stirring when all 4
        have passed.  For actual stir actions, the player must be current.
        """
        if self.state.phase != Phase.STIRRING:
            return False

        if stir is None:
            self._stir_passes.add(player_index)

            if player_index == self.state.current_player_index:
                self.state = self.state.model_copy(update={
                    "current_player_index": _next_player_sequential(player_index),
                })

            if len(self._stir_passes) >= PLAYER_COUNT:
                self.state = pickup_bottom_cards(self.state)

            return True

        # Actual stir: must be the current player
        if player_index != self.state.current_player_index:
            return False

        if not is_valid_stir(
            stir,
            self.state.trump_suit,
            self.state.trump_rank,
            self.state.stir_history,
        ):
            return False

        # Record stir manually (sequential turn order)
        self.state = self.state.model_copy(update={
            "trump_suit": stir.new_trump_suit,
            "trump_rank": stir.level or self.state.trump_rank,
            "stir_history": [*self.state.stir_history, stir],
            "current_player_index": _next_player_sequential(player_index),
        })
        self._stir_passes = set()
        return True

    # ---- Exchange (扣底) ----

    def submit_discard(self, player_index: int, cards: list[Card]) -> bool:
        """Declarer discards cards after picking up bottom cards."""
        if self.state.phase != Phase.EXCHANGE:
            return False
        if player_index != self.state.current_player_index:
            return False

        self.state = discard_cards(self.state, player_index, cards)
        return True

    # ---- Playing ----

    def submit_play(self, player_index: int, cards: list[Card]) -> bool:
        """Play cards from a player's hand.

        Determines the play type from legal plays when possible, otherwise
        infers the type from the cards themselves (SINGLE for 1 card, PAIR
        for 2 identical, etc.).  This keeps the method compatible with tests
        that play hand[0] without checking legality first.
        """
        if self.state.phase != Phase.PLAYING:
            return False
        if player_index != self.state.current_player_index:
            return False

        if not cards:
            return False

        # Try to match against legal plays first
        legal_plays = self._get_legal_plays(player_index)
        action = self._match_play_action(cards, legal_plays)

        # If no legal match, infer the play type from the cards
        if action is None:
            action = self._infer_play_type(cards)

        self.state = play_cards(self.state, player_index, action)
        if player_index == HUMAN_PLAYER_INDEX:
            self._ai_auto_play()
        return True

    def clear_trick(self) -> None:
        """Reset current trick slots for the next trick."""
        self.state = state_clear_trick(self.state)
        self.state = self.state.model_copy(
            update={"last_completed_trick": None}
        )

    # ---- Scoring / Next Round ----

    def next_round(self) -> None:
        """Calculate scores and advance to the next round."""
        if self.state.phase != Phase.SCORING:
            return

        result = self._calculate_round_score()

        if is_game_over(result.team0_new_level) or is_game_over(
            result.team1_new_level
        ):
            self.state = self.state.model_copy(
                update={"phase": Phase.GAME_OVER}
            )
            return

        new_declarer = (
            self.state.declarer_team_index
            if not result.switch_declarer
            else 1 - self.state.declarer_team_index
        )

        self.state = advance_round(
            self.state,
            result.team0_new_level,
            result.team1_new_level,
            new_declarer,
        )

    # ---- Query ----

    def get_awaiting_action(self) -> str | None:
        """Return what action the game is waiting for, or None."""
        phase = self.state.phase
        if phase == Phase.BIDDING:
            winner = get_winning_bid(self.state.bidding_history)
            if winner is not None and winner.level is not None:
                if is_bidding_over(self.state.bidding_history, PLAYER_COUNT):
                    return "set_trump"
            return "bid"
        if phase == Phase.STIRRING:
            return "stir"
        if phase == Phase.EXCHANGE:
            return "discard"
        if phase == Phase.PLAYING:
            trick = self.state.current_trick
            all_played = all(s.cards is not None for s in trick)
            if all_played or self.state.last_completed_trick is not None:
                all_cards_played = all(
                    len(p.hand) == 0 for p in self.state.players
                )
                if all_cards_played:
                    return "next_round"
                if self.state.last_completed_trick is not None:
                    return "clear_trick"
            return "play"
        if phase == Phase.SCORING:
            return "next_round"
        return None

    def is_human_turn(self) -> bool:
        """Check if it's the human player's turn."""
        return self.state.current_player_index == HUMAN_PLAYER_INDEX

    # ---- Private Helpers ----

    def _ai_auto_play(self) -> None:
        """Execute AI turns until human's turn or phase change."""
        from server.ai.auto_play import choose_bid, choose_discard, choose_play, choose_stir

        while not self.is_human_turn():
            phase = self.state.phase
            if phase not in (
                Phase.BIDDING, Phase.STIRRING, Phase.EXCHANGE, Phase.PLAYING
            ):
                break

            cp = self.state.current_player_index

            if phase == Phase.BIDDING:
                # Check if bidding already resolved
                if is_bidding_over(self.state.bidding_history, PLAYER_COUNT):
                    winner = get_winning_bid(self.state.bidding_history)
                    if winner is not None and winner.level is not None:
                        self.state = self.state.model_copy(
                            update={"current_player_index": winner.player_index}
                        )
                    break
                valid_levels = self.get_valid_bids()
                if not valid_levels:
                    break
                chosen = choose_bid(valid_levels, self.state.current_level)
                if chosen is None:
                    self.submit_bid(cp, None, pass_=True)
                else:
                    self.submit_bid(cp, chosen, pass_=False)

            elif phase == Phase.STIRRING:
                if self.state.trump_suit is None:
                    break
                valid = get_valid_stir_options(
                    self.state.trump_suit,
                    self.state.trump_rank,
                    cp,
                    self.state.stir_history,
                )
                if not valid:
                    self.submit_stir(cp, None)
                else:
                    result = choose_stir(
                        self.state.trump_suit,
                        [s.level for s in valid if s.level is not None],
                        cp,
                        self.state.stir_history,
                    )
                    if result is None:
                        self.submit_stir(cp, None)
                    else:
                        new_suit, level = result
                        stir = StirAction(
                            player_index=cp,
                            new_trump_suit=new_suit,
                            level=level,
                        )
                        self.submit_stir(cp, stir)

            elif phase == Phase.EXCHANGE:
                hand = self.state.players[cp].hand
                count = self.state.settings.bottom_card_count
                if len(hand) < count:
                    break
                discards = choose_discard(hand, count)
                self.submit_discard(cp, discards)

            elif phase == Phase.PLAYING:
                legal = self._get_legal_plays(cp)
                if not legal:
                    break
                action = choose_play(legal)
                self.submit_play(cp, action.cards)

    def _get_legal_plays(self, player_index: int) -> list[PlayAction]:
        """Get all legal plays for a player."""
        player = self.state.players[player_index]
        trick = self.state.current_trick
        is_leading = all(s.cards is None for s in trick)

        trick_dicts = [
            {"player_index": s.player_index, "cards": s.cards}
            for s in trick
        ]

        lead_action: PlayAction | None = None
        if not is_leading:
            for slot in trick:
                if slot.cards is not None:
                    lead_action = PlayAction(
                        type=self.state.lead_play_type or PlayType.SINGLE,
                        cards=slot.cards,
                    )
                    break

        return _get_legal_plays(
            player.hand,
            trick_dicts,
            self.state.trump_suit,
            self.state.trump_rank,
            is_leading,
            lead_action,
        )

    def _match_play_action(
        self, cards: list[Card], legal_plays: list[PlayAction]
    ) -> PlayAction | None:
        """Match played cards to a legal play action."""
        card_ids = {c.id for c in cards}
        for play in legal_plays:
            if card_ids == {c.id for c in play.cards}:
                return play
        return None

    @staticmethod
    def _infer_play_type(cards: list[Card]) -> PlayAction:
        """Infer play type from cards when no legal play matches."""
        if len(cards) == 1:
            return PlayAction(type=PlayType.SINGLE, cards=cards)
        if len(cards) == 2 and cards[0].suit == cards[1].suit and cards[0].rank == cards[1].rank:
            return PlayAction(type=PlayType.PAIR, cards=cards)
        # Default to SINGLE for any other combination
        return PlayAction(type=PlayType.SINGLE, cards=cards)

    def _calculate_round_score(self) -> ScoreResult:
        """Calculate the score for the current round."""
        last_trick = self.state.trick_history[-1]
        last_trick_winner_team = get_team_index(last_trick.winner_index)

        return calculate_score(
            self.state.defender_points,
            self.state.bottom_cards,
            last_trick_winner_team,
            last_trick.lead_type,
            self.state.declarer_team_index,
            self.state.teams[self.state.declarer_team_index].current_level,
            self.state.teams[1 - self.state.declarer_team_index].current_level,
        )
